From 872ca6e1958606e6ce9afc23c6041b972de316cb Mon Sep 17 00:00:00 2001
From: Bo Jiao <Bo.Jiao@mediatek.com>
Date: Fri, 16 May 2025 00:40:16 +0800
Subject: [PATCH 85/95] mtk: mt76: mt7996: fix the long registration time issue

Finish initializing VOW before starting the init work; otherwise, the
VOW command will be blocked for a long time, causing long boottime
issue. The result shows that boottime saves 9 seconds after rearranging
the functions.

---
 mt7996/init.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/mt7996/init.c b/mt7996/init.c
index c7e31cc4..6c633f42 100644
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -1972,9 +1972,6 @@ int mt7996_register_device(struct mt7996_dev *dev)
 	}
 
 	dev->mt76.token_threshold = MT7996_PER_BAND_TOKEN_SIZE;
-
-	ieee80211_queue_work(mt76_hw(dev), &dev->init_work);
-
 	dev->recovery.hw_init_done = true;
 
 	if (mt7996_vow_should_enable(dev)) {
@@ -1983,6 +1980,8 @@ int mt7996_register_device(struct mt7996_dev *dev)
 			goto error;
 	}
 
+	ieee80211_queue_work(mt76_hw(dev), &dev->init_work);
+
 	ret = mt7996_coredump_register(dev);
 	if (ret)
 		goto error;
-- 
2.45.2

